core/src/test/scala/unit/kafka/server/FetchRequestTest.scala

```diff
@@ -169,6 +171,58 @@ class FetchRequestTest extends BaseRequestTest {
   }
 
   /**
+   * Tests that down-conversions dont leak memory. Large down conversions are triggered
+   * in the server where the client closes connection. If references to byte arrays are
+   * not released this will OOM.
+   */
+  @Test
+  def testDownConversionWithConnectionFailure(): Unit = {
+    val (topicPartition, leaderId) = createTopics(numTopics = 1, numPartitions = 1).head
+
+    val producer = TestUtils.createNewProducer(TestUtils.getBrokerListStrFromServers(servers),
+      retries = 5, keySerializer = new StringSerializer, valueSerializer = new ByteArraySerializer)
+    try {
+      val bytes = new Array[Byte](100 * 1000)
+      for (i <- 0 to 3000) {
+        producer.send(new ProducerRecord(topicPartition.topic, topicPartition.partition,
+          "key", bytes))
+      }
+      producer.flush()
+    } finally {
+      producer.close()
+    }
+
+    def fetch(version: Short, waitForResponse: Boolean): Option[FetchResponse] = {
+      val maxPartitionBytes = 100 * 1024 * 1024
+      val fetchRequest = FetchRequest.Builder.forConsumer(Int.MaxValue, 0, createPartitionMap(maxPartitionBytes,
+        Seq(topicPartition))).build(version)
+
+      val socket = connect(brokerSocketServer(leaderId))
+      try {
+        val header = nextRequestHeader(ApiKeys.FETCH, fetchRequest.version)
+        val serializedBytes = fetchRequest.serialize(header).array
+        sendRequest(socket, serializedBytes)
+        if (waitForResponse) {
+          val response = receiveResponse(socket)
+          Some(FetchResponse.parse(skipResponseHeader(response), version))
+        } else {
+          // read some data to ensure broker has muted this channel
+          new DataInputStream(socket.getInputStream).readInt()
+          None
+        }
+      } finally {
+        socket.close()
+      }
+    }
+
+    (0 to 30).foreach(_ => fetch(1, false))
```

Comment:
Nit: can we use named arguments here and the other invocation to make the code clearer.